### ISSUE TITLE
Provide a logging hook to add custom labels to log entries

### ISSUE THREAD

**Issue by henkmollema** [GoogleCloudPlatform/google-cloud-dotnet#1937]
It would be useful if the diagnostics library provided a hook to add custom labels to log entries. Application Insights, for example, has a concept of [telemetry initializers](https://docs.microsoft.com/en-us/azure/application-insights/app-insights-api-filtering-sampling#add-properties-itelemetryinitializer) which allows you to add additional properties to telemetry.

Currently you can use `LoggerOptions.Labels` to provide custom labels:
https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/d0c81fd0756e69b664610b4438f4240f42d0f005/apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/LoggerOptions.cs#L42

However these labels are initialized once when configuring the logger. It would be nice if the `GoogleLogger` accepted a set of initializers to further populate these labels.

**Comment by iantalarico** [comment]
@henkmollema I'm assigning this to myself for now as I'd be the one to look into this more.  I'm a little low on bandwidth so this probably wouldn't happen for some time.  However, if you had a suggested design followed by PR I could manage that right now :)

**Comment by henkmollema** [comment]
I'll see if I can put something together :smile:

**Comment by henkmollema** [comment]
@iantalarico I'm thinking about a mechanism where as a user you can define one or more "log entry label providers" (e.g. `ILogEntryLabelProvider`? naming is hard) which are invoked in `GoogleLogger.Log(...)`:

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/94161eb1a9230c71b94c6b1ab4cfc4cc0f53bc54/apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/GoogleLogger.cs#L86

These providers could be resolved using the container in `IHttpContextAccessor` which is already available in the class, even though the use of the static `HttpContextAccessorWrapper` is a little bit tricky IMO.

The interface for a provider would look like this:

```cs
public interface ILogEntryLabelProvider
{
    void Invoke(Dictionary<string, string> labels);
}
```

I'll put up a design PR shortly. What are your thoughts about this?

**Comment by iantalarico** [comment]
@henkmollema 

I think the general approach is okay (and yes naming is hard but that name looks good).

The issue I see with this is ```IHttpContextAccessor``` from ```HttpContextAccessorWrapper``` is only available if when tracing is enabled.  That means if a user is only using logs and they define a ```ILogEntryLabelProvider``` the labels won't be added.

I also agree the HttpContextAccessorWrapper is a bit tricky (and hacky IMO).  It was done this way so we could grab trace ids from http headers without any breaking API changes.  I also plan to change this model in the future.

**Comment by henkmollema** [comment]
Yeah, it's very hacky 😄 I've considered your argument as well. For the sake quickly getting a design together I used the `HttpContextAccessorWrapper`.

I think it would be better to pass an `IServiceProvider` instance to the logger as that enables this feature without being in the context of an HTTP request. Or some composable configuration object. That would cause breaking changes though, as you mentioned.

**Comment by iantalarico** [comment]
@henkmollema Yes ideally we should pass an IServiceProvider into the logger and that would solve most issues like this moving forward.

@jskeet Do you think it's worthwhile to do another major release to enable this? It makes me cringe a little to do this so soon.

**Comment by jskeet** [comment]
Let's have a VC about it - I don't think I have enough context right now. (But this moment I'm in a car park...)

**Comment by iantalarico** [comment]
@jskeet Ping me if you have a few minutes to sync today or tomorrow.

**Comment by iantalarico** [comment]
@henkmollema After some more thought and discussion with @jskeet I think we are okay doing a major version bump for this.

The last hurdle I am seeing is ensuring that we support the[ recommended way of adding the providers in 2.x code](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?tabs=aspnetcore2x#how-to-add-providers).  Thoughts?

**Comment by henkmollema** [comment]
@iantalarico awesome! Do you want to consolidate all these changes into one PR or split them up? E.g. should we start with refactoring the `HttpContextAccessorWrapper` to an `IServiceProvider`.

We could add new extensions methods to the [`ILoggingBuilder`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.iloggingbuilder?view=aspnetcore-2.0) interface. At least for ASP.NET Core 2.0. I guess you want to keep support for ASP.NET Core 1.x as well so we would need both ways to register the logging providers. I could look into this if you'd like, which raises the same question: is this part of one big PR or separate ones.

**Comment by iantalarico** [comment]
@henkmollema  I always would rather have smaller PRs but won't complain if it's too much work to split them up. 

Yes we want to keep support for 1.x and for 2.0.  I think adding a new extension to ```ILoggingBuilder``` works.  Are you thinking of passing the ```IServiceProvider``` then?  I'm not sure we can get it during the initialization.

**Comment by henkmollema** [comment]
@iantalarico alright. I'll try to keep them separate 👍 

I think passing an `IServiceProvider` in the `ConfigureLogging()` callback is not possible. You won't have access to the application services container as it is still being built. We could register a startup filter and inject an `ILoggerFactory` and configure the logging from there. This allows you to use DI services.

I was thinking it would be nice to provide a single extension method on `IWebHostBuilder` as well which configures all the Google diagnostics at once with sensible defaults. Today if you want tracing, logging and error handling you'll have to configure it at 3 different places which is pretty clumsy.

What do you think about an extension method such as `IWebHostBuilder.UseGoogleDiagnostics()` which configures:
- Google logger provider
- Google tracing and exception logging services
- Google tracing and exception logging middleware

**Comment by iantalarico** [comment]
@henkmollema The startup filter sounds like a good approach to me.

> Today if you want tracing, logging and error handling you'll have to configure it at 3 different places which is pretty clumsy.

100% agree.

> What do you think about an extension method such as IWebHostBuilder.UseGoogleDiagnostics()

Very much yes.  This have been on my TODO list for sometime. :)

**Comment by iantalarico** [comment]
@henkmollema I will release a new version of the Google.Cloud.Diagnostics.AspNetCore library Monday!

**Comment by henkmollema** [comment]
Awesome, thanks!

**Comment by iantalarico** [comment]
@henkmollema This is now released as [3.0.0-beta01](https://www.nuget.org/packages/Google.Cloud.Diagnostics.AspNetCore/3.0.0-beta01)

**Comment by henkmollema** [comment]
Thanks Ian! I'll try it out this week.
